timemachine/cpp/src/gpu_utils.cuh

```diff
@@ -83,3 +84,85 @@ double __device__ __forceinline__ rmul_rn(double a, double b) { return __dmul_rn
 float __device__ __forceinline__ radd_rn(float a, float b) { return __fadd_rn(a, b); }
 
 double __device__ __forceinline__ radd_rn(double a, double b) { return __dadd_rn(a, b); }
+
+// If this were not int128s, we could do __shfl_down_sync, but only supports up to 64 values
+__device__ __forceinline__ void energy_warp_reduce(volatile __int128 *shared_data, int thread_idx) {
+#pragma unroll 6
+    for (int i = 32; i > 0; i /= 2) {
+        if (thread_idx < i) {
+            shared_data[thread_idx] = shared_data[thread_idx] + shared_data[thread_idx + i];
+        }
+        __syncwarp();
+    }
+}
+
+template <unsigned int THREADS_PER_BLOCK>
+__device__ __forceinline__ void block_energy_reduce(volatile __int128 *shared_data, int tid) {
```

Comment:
Worth adding a comment here? (many "magic numbers", and not clear to me what's going on here)

Reply:
Added docstring here https://github.com/proteneer/timemachine/pull/1090/commits/b0dc781532902ab665cf012727add165fb56fd1f